### PR TITLE
Locale and timezone setting support

### PR DESCRIPTION
**Pull request by slakkala**
#

## Description

Implement locale and timezone setting methods in givc-admin, and report the settings back to agents upon registration. Implement locale and timezone setting application in givc-agent (using localectl and timedatectl).

## Checklist

- [X] Summary of the proposed changes in the PR description
- [X] More detailed description in the commit message(s)
- [X] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [ ] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Test procedure added to nixos/tests
- [X] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf-givc/actions)
- [ ] Author has added reviewers and removed PR draft status

### CODE REVIEW

**Review by mbssrc — APPROVED** [review]
Let's add a test case for this later

File: client/src/client.rs [hunk 1]
```diff
@@ -1,32 +1,23 @@
 use anyhow::bail;
 use async_channel::Receiver;
+use givc_common::pb;
+pub use givc_common::query::{Event, QueryResult};
 use tokio_stream::StreamExt;
 use tonic::transport::Channel;
 use tracing::debug;
 
 use givc_common::address::EndpointAddress;
-use givc_common::pb;
```
Comment (avnik) [inline client/src/client.rs:8]: I (started) grouped imports intentionally -- std, libraries, project crates, this crates import (internally alphabetically sorted)

File: nixos/modules/appvm.nix [hunk 2]
```diff
@@ -175,6 +175,21 @@ in
       }
     ];
 
+    security.polkit = {
+      enable = true;
+      extraConfig = ''
+        polkit.addRule(function(action, subject) {
+            if (action.id == "org.freedesktop.locale1.set-locale" && subject.user == "ghaf") {
```
Comment (avnik) [inline nixos/modules/appvm.nix:182]: Should username be hardcoded? Or we should have `ghaf.username`  setting somewhere)

File: client/src/client.rs [hunk 3]
```diff
@@ -1,32 +1,23 @@
...
 pub struct WatchResult {
     pub initial: Vec<QueryResult>,
     // Design defence: we use `async-channel` here, as it could be used with both
     // tokio's and glib's eventloop, and recommended by gtk4-rs developers:
     pub channel: Receiver<Event>,
-
-    task: tokio::task::JoinHandle<()>,
-}
-
-impl Drop for WatchResult {
```
Comment (avnik) [inline client/src/client.rs:26]: I am concerning about remove `Drop` here,   I not sure if task spawned for polling grpc would shut down automatically, if `WatchResult` is destroyed.

**Review by avnik — COMMENTED** [review]
Generally LGTM, I planning to approve it  as soon as you resolve some small questions.

Also next time please split feature PR and massive style refactoring PR (I do this mistake often myself as well)

File: client/src/client.rs (hunk 1 above)
Comment (slakkala) [reply client/src/client.rs:8]: Ah, yes, this got messed up in a rebase.

File: client/src/client.rs (hunk 3 above)
Comment (slakkala) [reply client/src/client.rs:26]: Without the Drop implementation, the spawned task will run until there is an event, at which point the `recv` will fail and task will terminate.

This was mainly for larger refactoring for helping interoperability with gtk-rs, but turned out it actually made things worse.

My main concern with this is that the `WatchResult` cannot be destructured, but I guess we can live with that.

File: nixos/modules/appvm.nix (hunk 2 above)
Comment (slakkala) [reply nixos/modules/appvm.nix:182]: Good question; I'll check.

**Comment by slakkala** [comment]
> Also next time please split feature PR and massive style refactoring PR (I do this mistake often myself as well)

I did intentionally put them in separate commits, but yeah, maybe PRs would've been smarter.

File: client/src/client.rs (hunk 3 above)
Comment (slakkala) [reply client/src/client.rs:26]: Changed it now to use a tokio mpsc channel for signalling; no need to explicitly implement Drop: when the (last, but in this case only) sender is dropped, the `recv()` -future will complete, and the task will terminate.

File: nixos/modules/appvm.nix (hunk 2 above)
Comment (slakkala) [reply nixos/modules/appvm.nix]: Better now?

Comment (slakkala) [reply nixos/modules/appvm.nix]: That broke standalone things, reverting for now.

**Pull request merged**